apis/apiextensions/v1/composition_common.go

```diff
@@ -117,3 +124,42 @@ const (
 	// credentials from a secret.
 	FunctionCredentialsSourceSecret FunctionCredentialsSource = "Secret"
 )
+
+// FunctionRequirements define requirements that a function may need to
+// satisfy.
+type FunctionRequirements struct {
+	// RequiredResources is a list of resources that must be fetched before
+	// this function is called.
+	// +optional
+	// +listType=map
+	// +listMapKey=requirementName
+	RequiredResources []RequiredResourceSelector `json:"requiredResources,omitempty"`
+}
+
+// RequiredResourceSelector selects a required resource.
+//
+// +kubebuilder:validation:XValidation:rule="(has(self.name) && !has(self.matchLabels)) || (!has(self.name) && has(self.matchLabels))",message="Either name or matchLabels must be specified, but not both"
+type RequiredResourceSelector struct {
```

Comment:
i think we talked about this in the first Operations PR, but it looks like this type is duplicated in operations common types too - worth de-duplicating these, or not because ops are v1alpha still?

Reply:
Yeah I'd rather wait until ops are GA - then maybe we could have a shared pipeline types dir.

Though I don't really mind duplicating them TBH. Loose coupling has benefits if we need one to diverge from the other. It also has risks though -  i.e. we update one but don't update the other.